chore(iOS): deprecate RCTAppDelegate

okwasniewski opened the pull request:
## Summary:


Recently, I've introduced `RCTReactNativeFactory` in this PR: #46298, which is a good successor for `RCTAppDelegate`. 

### Why?

`RCTAppDelegate` introduced strong coupling between React Native and AppDelegate pattern. From iOS 13+ there is a newer equivalent (Scene Delegate) which is not possible to achieve with current architecture. The proposed solution involves migration to a `RCTReactNativeFactory` a class that encapsulates initialization logic of React Native. 

This migration will make brownfield initialization easier by making it more flexible and simpler to integrate into already established apps. 

### Deprecation plan

The plan I've discussed with @cipolleschi involves:

- Deprecation of `RCTAppDelegate` in 0.79 (current main)
- Migration off `RCTAppDelegate` to SceneDelegate + `RCTReactNativeFactory` in 0.80


## Changelog:

[IOS] [DEPRECATED] - deprecate RCTAppDelegate

## Test Plan:

Not needed

cipolleschi reviewed (changes requested):
@okwasniewski thanks for the PR. 

This change is definitely needed but I kind of expected it to comes together with the changes in RNTester and the Helloworld app. We can't just deprecate RCTAppDelegate without showing users how to properly use the `RCTReactNativeFactory`. Are you planning to submit them in a separate PR?

RyanCommits commented:
Noticed that in `packages/react-native/Libraries/AppDelegate/RCTReactNativeFactory.mm` we not longer have `enableFixForViewCommandRace`. 

Was this omitted on purpose in PR #46298? @okwasniewski 
```
bool enableFixForViewCommandRace() override
  {
    return true;
  }
  ```

okwasniewski commented:
@RyanCommits You are right! I missed this flag during refactor, created a follow-up PR: https://github.com/facebook/react-native/pull/49126

cipolleschi commented on packages/rn-tester/RNTester/AppDelegate.h:
```diff
@@ -5,9 +5,13 @@
+@interface AppDelegate : RCTDefaultReactNativeFactoryDelegate <UIApplicationDelegate, UISceneDelegate>
+
+@property(nonatomic, strong, nonnull) UIWindow *window;
+@property(nonatomic, strong, nonnull) RCTReactNativeFactory *reactNativeFactory;
```
maybe this should be `readonly`?

okwasniewski replied:
I need to write to both of these properties in AppDelegate: 

![CleanShot 2025-02-03 at 15 04 38@2x](https://github.com/user-attachments/assets/5463f781-f718-4c09-b426-17fbbd35c045)

Same with window

cipolleschi replied:
😭 We don't have a `private(set)` in objective-c, have we?

okwasniewski replied:
Unfortunately not 😢 We could work around this by overwriting setters but it's user's AppDelegate not API from RN so Im not sure if its worth doing

cipolleschi commented on packages/rn-tester/RNTester/NativeExampleViews/FlexibleSizeExampleView.mm:
```diff
@@ -47,7 +47,7 @@ - (instancetype)initWithFrame:(CGRect)frame
     AppDelegate *appDelegate = (AppDelegate *)[[UIApplication sharedApplication] delegate];
 
     _resizableRootView =
-        (RCTRootView *)[appDelegate.rootViewFactory viewWithModuleName:@"RootViewSizeFlexibilityExampleApp"];
+        (RCTRootView *)[appDelegate.reactNativeFactory.rootViewFactory viewWithModuleName:@"RootViewSizeFlexibilityExampleApp"];
```
I wonder if we can find a way not to reach for the AppDelegate... 🤔 but probably this is a problem for another time.

okwasniewski replied:
Probably users could implement their own singelton wrapper for the factory? Not sure if there is other way to handle it. Making it a singelton by default isn't the best option as users may want to spawn multiple instances.

cipolleschi reviewed (changes requested):
Thanks for updating the PR. I left some extra comments.

cipolleschi reviewed (changes requested):
Thanks for updating the PR. I left some extra comments.

cipolleschi reviewed (approved):
It looks good to me! Thanks for working on this.

Let import this and hope that it works well internally. (spoiler: I believe it won't... xD)

cipolleschi commented on packages/helloworld/ios/HelloWorld/AppDelegate.swift:
```diff
@@ -10,19 +10,49 @@ import React_RCTAppDelegate
+class AppDelegate: UIResponder, UIApplicationDelegate, UIWindowSceneDelegate {
+  var window: UIWindow?
+  
+  var reactNativeFactory: RCTReactNativeFactory!
+  var reactNativeDelegate: ReactNativeDelegate!
```
our swift lint rule don't really like implicitly unwrapped optionals. Can we add a couple of `// swiftlint:disable:next` comments?

cipolleschi commented on packages/helloworld/ios/HelloWorld/AppDelegate.swift:
```diff
@@ -10,19 +10,49 @@ import React_RCTAppDelegate
-    // You can add your custom initial props in the dictionary below.
-    // They will be passed down to the ViewController used by React Native.
-    self.initialProps = [:]
+    let rootView = reactNativeFactory.rootViewFactory.view(
+      withModuleName: "HelloWorld",
+      initialProperties: [:],
+      launchOptions: launchOptions
+    )
+    
+    window = UIWindow(frame: UIScreen.main.bounds)
+    window?.windowScene?.delegate = self
+    let rootViewController = UIViewController()
+    rootViewController.view = rootView
+    window?.rootViewController = rootViewController
+    window?.makeKeyAndVisible()
```
Yesterday evening I was think about this bit an I feel like this is too verbose. 
I'm a bit worried that most people will have a hard time migrating, so I thought we might simplify it.

What do you think about adding this method to RCTReactNativeFactory: 
```swift
function startReactNative(module moduleName: String, in window: UIWindow?, with launchOptions: launchOptions) {
  let rootView = reactNativeFactory.rootViewFactory.view(
      withModuleName: moduleName,
      initialProperties: [:],
      launchOptions: launchOptions
    )
  let rootViewController = UIViewController()
  rootViewController.view = rootView
  window?.rootViewController = rootViewController
  window?.makeKeyAndVisible()
}
```

In this way, the default way to launch React Native is wrapped in the factory and the code change required to migrate is more contained.

```swift
reactNativeDelegate = ReactNativeDelegate()
reactNativeFactory = RCTReactNativeFactory(delegate: reactNativeDelegat)
reactNativeDelegate.dependencyProvider = RCTAppDependencyProvider()
window = UIWindow(frame: UIScreen.main.bounds)
window?.windowScene?.delegate = self
reactNativeFactory.startReactNative(module: "HelloWorld", in: window, with: launchOptions)
return true
```

I feel that this conveys better what we are trying to do. And people that has more complex setups can just copy an paste the part of the code they need.

okwasniewski replied:
Yeah looks good! This will make migration to scene delegate smoother too

okwasniewski replied:
Added this new method, can you take a look?

cortinico commented:
> Recently, I've introduced `RCTReactNativeFactory` in this PR: #46298, which is a good successor for `RCTAppDelegate`.

Can you update the commit message/summary of this PR to explain **why** we're deprecating `RCTAppDelegate`, just mentioning that it's a good successor is sadly not enough. Otherwise folks will have to open your original PR to understand why this deprecation is happening.

okwasniewski commented:
> > Recently, I've introduced `RCTReactNativeFactory` in this PR: #46298, which is a good successor for `RCTAppDelegate`.
> 
> Can you update the commit message/summary of this PR to explain **why** we're deprecating `RCTAppDelegate`, just mentioning that it's a good successor is sadly not enough. Otherwise folks will have to open your original PR to understand why this deprecation is happening.

@cortinico I've updated description, can you check it?

cortinico commented:
> @cortinico I've updated description, can you check it?

LGTM 👍

cipolleschi reviewed (approved):
Much better, thank you for working on this and improving the happy path!
